ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/DB/HdWallet/Delete.hs

```diff
@@ -10,15 +12,38 @@ import Control.Lens (at, (.=))
 
 import Ariadne.Wallet.Cardano.Kernel.DB.HdWallet
 import Ariadne.Wallet.Cardano.Kernel.DB.Util.AcidState
+import Ariadne.Wallet.Cardano.Kernel.DB.HdWallet.Read
 
 {-------------------------------------------------------------------------------
   DELETE
 -------------------------------------------------------------------------------}
 
--- | Delete a wallet
-deleteHdRoot :: HdRootId -> Update' HdWallets e ()
-deleteHdRoot rootId = zoom hdWalletsRoots $ at rootId .= Nothing
+-- | Delete a wallet with the whole subtree (addresses and accounts)
+deleteHdRoot :: HdWallets -> HdRootId -> Update' HdWallets e ()
+deleteHdRoot hdwallets rootId = do 
+  let 
+     (accs,addrs) = getAccountsAndAddresses hdwallets rootId
+     accIds  = accs <&> (^. hdAccountId)
+     addrIds = addrs <&> (^. hdAddressId)
+  zoom hdWalletsAddresses $ mapM_ (\addrId -> (at addrId .= Nothing)) addrIds
+  zoom hdWalletsAccounts $ mapM_ (\accId -> (at accId .= Nothing)) accIds
+  zoom hdWalletsRoots $ at rootId .= Nothing
+  where 
+    getAccountsAndAddresses :: HdWallets -> HdRootId -> ([HdAccount],[HdAddress])
+    getAccountsAndAddresses wallets rootID = 
+      fromRight (error "Bug: UnknownHdRoot") $ 
+      (bimap toList toList) <$>
+      ((,) <$> (readAccountsByRootId rootID wallets) <*> (readAddressesByRootId rootId wallets))
+
 
--- | Delete an account
-deleteHdAccount :: HdAccountId -> Update' HdWallets UnknownHdRoot ()
-deleteHdAccount accId = zoom hdWalletsAccounts $ at accId .= Nothing
+-- | Delete an account with it's addresses
+deleteHdAccount :: HdWallets -> HdAccountId -> Update' HdWallets UnknownHdRoot ()
+deleteHdAccount hdwallets accId = do 
+  let addrIds = (getAddressesByAccount hdwallets accId) <&> (^. hdAddressId)
+  zoom hdWalletsAddresses $ mapM_ (\addrId -> (at addrId .= Nothing)) addrIds
```

Comment:
Here and in `deleteHdRoot` we basically want to delete all items from `IxSet` with given index. So I have a cosmetic suggestion: define a function with type `(Indexable a, IsIndexOf ix a) => ix -> IxSet a -> IxSet a` in `Ariadne.Wallet.Cardano.Kernel.DB.Util.IxSet` and use it here and above. It allows to avoid usage of `error` and reduces code duplication.
Possible definition (I didn't test it):
```
deleteIxAll ix (WrapIxSet ixSet) =
    WrapIxSet (IxSet.getGT ix ixSet `IxSet.union` IxSet.getLT ix ixSet)
```
Note: there is `deleteIx` in `ixset-typed`, but it can't be used, because it
> Will not change IxSet if you have more than one item with given index.